### ISSUE TITLE
Occasional crash from DynamicSoundEffectInstance

### ISSUE THREAD

**Issue by vrgamestudio** [MonoGame/MonoGame#8064]
Very occasionally I see a random program crash occurring from DynamicSoundEffectInstance (XAudio Platform):

```
System.ArgumentException: Source array was not long enough. Check the source index, length, and the array's lower bounds. (Parameter 'sourceArray')
   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length)
   at System.Collections.Generic.Queue`1.SetCapacity(Int32 capacity)
   at System.Collections.Generic.Queue`1.Grow(Int32 capacity)
   at Microsoft.Xna.Framework.Audio.DynamicSoundEffectInstance.PlatformSubmitBuffer(Byte[] buffer, Int32 offset, Int32 count) in D:\Projects\Library\VRGameStudio-MonogameSource\MonoGame\MonoGame.Framework\Platform\Audio\DynamicSoundEffectInstance.XAudio.cs:line 78
```

From what I've been able to gather so far this is likely due to DynamicSoundEffectInstance.XAudio not using a safe threaded Queue?
There is a ConcurrentQueue class that maybe the answer but I'm wondering if anyone has any experience of this, what exactly is causing it or could recommend any approaches to tackle this (very occasional, hardly affects anyone ever, but if it does it will crash your game) bug?

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8.1

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- WindowsDX, WindowsUWP

**Comment by dellis1972** [comment]
@vrgamestudio are you able to provide some sample code of what you are doing so we can reproduce the issue?

**Comment by vrgamestudio** [comment]
I'm afraid not since it's one of those issues that would only crop up very occasionally during extensive testing of a large project but it was a serious problem that would crash them game if it happened. 

However I did find a workaround as mentioned above using the ConcurrentQueue class. Since implementing that I've not noticed any bad things happen and the bug has not appeared since. Basically I:

Changed the definitions of this:
```
    public sealed partial class DynamicSoundEffectInstance : SoundEffectInstance
    {
        private Queue<AudioBuffer> _queuedBuffers;
        private Queue<byte[]> _pooledBuffers;
        ...
```


to this:
```
    public sealed partial class DynamicSoundEffectInstance : SoundEffectInstance
    {
        private ConcurrentQueue<AudioBuffer> _queuedBuffers;
        private ConcurrentQueue<byte[]> _pooledBuffers;
        ...
```

**Comment by mrhelmut** [comment]
Interesting. Are you feeding the ```DynamicSoundEffectInstance``` from a thread?

**Comment by vrgamestudio** [comment]
No but...(some extra detail that may not be directly relevant) the creation of the object is done from a Task (to speed up Game Startup time during Initialization/LoadContent). So basically this runs in a Task (in a child IGameComponent):

```
public override void Initialize()
{
    GameSounds = new DynamicSoundEffectInstance(44100, AudioChannels.Stereo) { Volume = 1.0f };
    ...
```

This did cause an interesting problem (but only in UWP), in that it would crash UWP because it turned out that the SoundEffect class was Initialized on the "wrong Thread for UWP land". Basically the above call to the DynamicSoundEffectInstance constructor will make a call to SoundEffect.Initialize() (so calling it from a thread). This would crash like this during startup (in Program.cs):

![SoundEffect Initialized from Thread](https://github.com/MonoGame/MonoGame/assets/74532136/717d96ee-ba55-4576-a362-f503cbb8107f)

There is no useful Call Stack or details that help at all. Such an error did not happen on Desktop DX and was a complete surprise and headscratcher at the time.

So to solve for UWP but still be compatible with other platforms, I manually call SoundEffect.Initialize() from  the main thread during Game.Initialize():

```
protected override void Initialize()
{
    // ensure that soundeffect.initialize is called from the main thread (otherwise causes an issue with uwp on pc/Xbox).
    SoundEffect.Initialize();
    ...
```

So having said all that, ultimately byte samples are sent during the main thread's Draw() call:

```
GameSounds.SubmitBuffer(samples);

if (GameSounds.State != SoundState.Playing)
    GameSounds.Play();
```

Hence the answer to your question is no.

To be clear; the problem of the crash from DynamicSoundEffectInstance seemed to occur in both UWP and Desktop DX IIRC (although mostly on Desktop DX since that's what I tested with most for build time reasons) and would happen maybe once or twice a day during a long coding/testing session, totally at random. My feeling was that this was a problem across the board, potentially with all platforms...I was just concentrating on Desktop DX and UWP.

The problem with the UWP initialization stated above is not really anything to do with it AFAIK, just a quirk I (eventually) found, however I suspect there is some internal thread code (in MonoGame or SharpDX?) that means that these Queues are not always updated by the same thread. I have no proof, that was just a hunch which led to my hacky solution above and which does seem to have cured the problem totally, as in it's never happened since, not even once. It's possible it introduces a side effect by way of a thread block while waiting for access but I've not encountered anything yet that would indicate there is a problem with it.

Before I found the ConcurrentQueue work around I had hacked around with DynamicSoundEffectInstance.PlatformDispose and DynamicSoundEffectInstance.OnBufferEnd (the latter is an event handler from SourceVoice.BufferEnd), which seemed at the time to be where the problem was coming from. I ended up with code like this (trying to take care with null values). From something like this:

```
////Original code...
//var buffer = _queuedBuffers.Dequeue();
//buffer.Stream.Dispose();
//_bufferPool.Return(_pooledBuffers.Dequeue());
```

To this:

```
// New version of ORIGINAL code above...
AudioBuffer buffer;
if (_queuedBuffers.TryDequeue(out buffer))
{
    buffer.Stream.Dispose();
    byte[] bytes = null;
    if (_pooledBuffers.TryDequeue(out bytes))
        _bufferPool.Return(bytes);
}
```

And at times I would talk myself into it, "Ah that has solved it"...only for the error to return much later. I still have this more careful code in my version but I don't think it does anything useful as I was still getting the error and it didn't really solve anything.

**Comment by dellis1972** [comment]
@vrgamestudio 

When you say task do you mean a `System.Threading.Task` ? i.e `var task = new Task (() => { // Do my stuff });` kinda thing?
If so that will be using the `ThreadPool` in the background. So it won't be running on the main thread.

**Comment by vrgamestudio** [comment]
Yes correct I do mean that. It was a long post though so do bear in mind that was only for the creation of the object and;

`GameSounds.SubmitBuffer(samples);`

only happens from the main thread during Draw().

Whether it makes any difference if an object is created from one thread and later used from another, I'm not sure, I don't think so as long as the object is good and initialized before you start using it? If the problem arose at all it was long, long after initialization, sometimes hours of gameplay later.

Regardless though I've heard it said that pumping samples in from another thread can also be a good idea, at least back then:

[https://twitter.com/Rolweg91/status/29041459253](https://twitter.com/Rolweg91/status/29041459253)

So it might well be that to use a thread safe Queue is just a better idea all round. Whether or not there is MonoGame or SharpDX code that uses separate threads to interact with those Queues (my guess is yes, at least on Desktop DX and UWP) or if the API consumer wants to use threading for performance or other reasons, then every which way it's all covered?

**Comment by mrhelmut** [comment]
Yeah, UWP is very sensitive in regard to what is executed on the UI thread, most things that call to system stuff (like sound) won't be working outside the UI thread. Moreover, XAudio2 (the underlying audio engine used by both UWP and WindowsDX) doesn't like operations to be made from different thread (it is fine with threads, but it has a strong preference toward having all operations on the same thread than the one it has been initialized on).

MonoGame is not designed as thread-safe. We're trying to guarantee that loading content is thread-safe, but that's it. Though "not thread-safe" doesn't mean that it can't work with threads, but that thread-safety is left to the user to handle. We should try to note in the documentation what is thread-safe and what isn't, and suggest users to make their own code thread-safe (which is a better practice than teaching to expect that third-party code is thread-safe).

----

Calling ```SoundEffect.Initialize();``` manually from the main thread _before_ initializing any sound effect is the way to go whenever your first sound effect initialization is done from a thread (that's when audio is internally initialized). 👍 Manually calling it in advance forces MonoGame to init audio on the right thread. I believe that we need to highlight this in a threading documentation. But we can also improve this by having a ```BlockOnUIThread``` inside ```SoundEffect.Initialize();```.

Feeding instances with ```SubmitBuffers()``` from a thread is fine but... users should add locks around any calls to ```SubmutBuffers()``` (actually any audio operations).

I don't think that ```ConcurrentQueue``` is the solution here. I feel like it's a patch over what might be insufficient thread-safety in user code and might become the tree that hides the forest.

TL;DR; we need to identify exactly what is happening before adding threading locks anywhere.

**Comment by vrgamestudio** [comment]
> Feeding instances with SubmitBuffers() from a thread is fine but... users should add locks around any calls to SubmutBuffers() (actually any audio operations).

Do you mean something like this (when submitting samples from another thread)?:

```
lock (GameSounds)
{
    GameSounds.SubmitBuffer(samples);

    if (GameSounds.State != SoundState.Playing)
        GameSounds.Play();
}
```
But what about when not submitting samples from another thread? In my case I'm doing it during the Draw() method, from the main thread.

Is it because I may have instantiated the "GameSounds" object on another thread and UWP is very particular? Well then the error happens on Desktop DX also, at the least. Maybe Desktop DX is very particular too, I don't know.

I feel so dumb like I'm missing something here. Why would I even think I needed a lock in this case?

**Comment by mrhelmut** [comment]
> Is it because I may have instantiated the "GameSounds" object on another thread and UWP is very particular? Well then the error happens on Desktop DX also, at the least. Maybe Desktop DX is very particular too, I don't know.

They are both using XAudio 2 as their audio engine, so they're going to behave exactly the same sound-wise. UWP has some additional sensitivity in regard to threading which are related to its underlying WinRT runtime (it doesn't like some stuff to be initialized from a thread). Moreover, the best practice when XAudio is involved, is to have all your audio calls from the same thread.
Long story short: if you initialized the audio on a specific thread, all your audio calls should be running on that thread.


Could you share a minimal project exposing your issue? I feel like the actual issue is not related to threading and that adding ConcurrentQueue made us think it is but is hiding something else.

**Comment by vrgamestudio** [comment]
I didn't think it would be possible but I've managed to put together the simplest example I can that demonstrates what I think was the problem I was seeing.

The only real fakeness of it is the frequency of the sound sample data (submitting bytes every single frame) and the fact it runs with an unlocked framerate just to make it fail as fast as possible. 

[DynamicSoundTest.zip](https://github.com/MonoGame/MonoGame/files/13934097/DynamicSoundTest.zip)

It should be self explanatory, it's for Visual Studio 2022 and there are comments in the code.

From memory these were the original errors I used to get before I started messing around with it. The two main errors I am seeing with this program are:
![ErrorByteBufferPool](https://github.com/MonoGame/MonoGame/assets/74532136/75d73bd4-86e7-4027-8168-f35c3caaa390)
![ErrorInQueue](https://github.com/MonoGame/MonoGame/assets/74532136/e9113231-04de-4e2a-a7a7-15548e163b3c)

**Comment by mrhelmut** [comment]
I don't see any bug in your sample. What I see is that you're feeding buffers that are longer than the duration of a frame (providing that the sample runs fast enough, which it does on my computer) at each frame. Which means that you are overloading the buffer pool until it runs out of memory because buffers don't have the time to be processed. You're creating a traffic jam.

You shouldn't submit that many buffers if they are longer than the duration of a frame. You can have a few ones submitted ahead of time to avoid any cuts, but certainly not submitting thousands ahead of time. I would suggest limiting the number of buffers and feeding them only when the sound instance is about to run out of stuff to play (you can use ```DynamicSoundEffectInstance.PendingBufferCount``` and feeding new buffers only when the pending count if below 2 or 3).

If what you are trying to do is sampling a microphone and sending that to a sound instance, there must be something done wrong. You might want to check if your buffer is ```Microphone.GetSampleSizeInBytes(Microphone.BufferDuration)``` long and that you are sending data to sound instance only when ```Microphone.BufferReady``` is triggered. Not on every frame.

**Comment by vrgamestudio** [comment]
It's lovely for you that you didn't encounter any bug but I'm sure you'll agree that doesn't necessarily mean that there isn't one.

Did you play with the size of the _soundBuffer array to see at which point you *would* see an error? I would guess that at some point you would have?

I realise I was overloading the buffer, it was a test after all, "what happens if you simulate ever so slightly overloading the buffer over a long period of time but do it much more quickly?" Of course in the real world I would not send this much data so quickly but I did so here to attempt to simulate a problem that may only crop up after several hours play. With the best will in the world, however carefully one may try to time the number of samples to the framerate, it may well end up being a little too much (after all we want to avoid sound stutters so we definitely don't want too FEW samples), which over time could build up.

However, rather than an out of memory exception as your description would suggest, on my machine I was getting a variety of other exceptions as you would see above.

So OK given your advice, lets play it safe and I changed the code of my test to this:

```
if (_gameSound.PendingBufferCount < 3)
{
    _gameSound.SubmitBuffer(_soundBuffer);
    if (_gameSound.State != SoundState.Playing)
        _gameSound.Play();
}

```
And it gives the exact same result, an exception is thrown in the Queue class quite quickly.

So next I tried this:

```
if (_gameSound.PendingBufferCount < 2)
{
    _gameSound.SubmitBuffer(_soundBuffer);
    if (_gameSound.State != SoundState.Playing)
        _gameSound.Play();
}
```

And it lasted a little longer, 3 min 44 seconds but again it fell over.

So guess what I did next? Yup:

```
if (_gameSound.PendingBufferCount < 1)
{
    _gameSound.SubmitBuffer(_soundBuffer);
    if (_gameSound.State != SoundState.Playing)
        _gameSound.Play();
}
```
Can you guess what happened? Yes, the exact same result.

So I'm sorry you weren't able to replicate the bug on you machine but I'm pretty sure there is a problem regardless. I suggest you play with the size of the _sampleBuffer array to confirm this?

Just to round this off I took this program and REMOVED the line:

`if (_gameSound.PendingBufferCount < 1)`

So there was no "buffer protection" (which didn't work anyway) and ran it against the latest source code on Git. It was no surprise that it was the same result, an exception of the types above, as expected.

So then I modified the file MonoGame.Framework.WindowsDX/Platform/Audio/DynamicSoundEffectInstance.XAudio.cs in the following way:

```
// MonoGame - Copyright (C) The MonoGame Team
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using MonoGame.Framework.Utilities;
using SharpDX;
using SharpDX.Multimedia;
using SharpDX.XAudio2;

namespace Microsoft.Xna.Framework.Audio
{
    public sealed partial class DynamicSoundEffectInstance : SoundEffectInstance
    {
        //private Queue<AudioBuffer> _queuedBuffers;
        //private Queue<byte[]> _pooledBuffers;
        private ConcurrentQueue<AudioBuffer> _queuedBuffers;
        private ConcurrentQueue<byte[]> _pooledBuffers;
        private static ByteBufferPool _bufferPool = new ByteBufferPool();

        private void PlatformCreate()
        {
            _format = new WaveFormat(_sampleRate, (int)_channels);
            _voice = new SourceVoice(SoundEffect.Device, _format, true);
            _voice.BufferEnd += OnBufferEnd;
            //_queuedBuffers = new Queue<AudioBuffer>();
            //_pooledBuffers = new Queue<byte[]>();
            _queuedBuffers = new ConcurrentQueue<AudioBuffer>();
            _pooledBuffers = new ConcurrentQueue<byte[]>();
        }

        private int PlatformGetPendingBufferCount()
        {
            return _queuedBuffers.Count;
        }

        private void PlatformPlay()
        {
            _voice.Start();
        }

        private void PlatformPause()
        {
            _voice.Stop();
        }

        private void PlatformResume()
        {
            _voice.Start();
        }

        private void PlatformStop()
        {
            _voice.Stop();

            // Dequeue all the submitted buffers
            _voice.FlushSourceBuffers();
        }

        private void PlatformSubmitBuffer(byte[] buffer, int offset, int count)
        {
            // we need to copy so datastream does not pin the buffer that the user might modify later
            byte[] pooledBuffer;
            pooledBuffer = _bufferPool.Get(count);
            _pooledBuffers.Enqueue(pooledBuffer);
            Buffer.BlockCopy(buffer, offset, pooledBuffer, 0, count);

            var stream = DataStream.Create(pooledBuffer, true, false, 0, true);
            var audioBuffer = new AudioBuffer(stream);
            audioBuffer.AudioBytes = count;

            _voice.SubmitSourceBuffer(audioBuffer, null);
            _queuedBuffers.Enqueue(audioBuffer);
        }

        private void PlatformUpdateQueue()
        {
            // The XAudio implementation utilizes callbacks, so no work here.
        }

        private void PlatformDispose(bool disposing)
        {
            if (disposing)
            {
                while (_queuedBuffers.Count > 0)
                {
                    //var buffer = _queuedBuffers.Dequeue();
                    //buffer.Stream.Dispose();
                    //_bufferPool.Return(_pooledBuffers.Dequeue());

                    //var buffer = _queuedBuffers.Dequeue();
                    AudioBuffer buffer;
                    if (_queuedBuffers.TryDequeue(out buffer))
                    {
                        buffer.Stream.Dispose();
                        byte[] bytes = null;
                        if (_pooledBuffers.TryDequeue(out bytes))
                            _bufferPool.Return(bytes);
                    }
                }
            }
            // _voice is disposed by SoundEffectInstance.PlatformDispose
        }

        private void OnBufferEnd(IntPtr obj)
        {
            // Release the buffer
            if (_queuedBuffers.Count > 0)
            {
                //var buffer = _queuedBuffers.Dequeue();
                //buffer.Stream.Dispose();
                //_bufferPool.Return(_pooledBuffers.Dequeue());

                // New version of ORIGINAL code above...
                AudioBuffer buffer;
                if (_queuedBuffers.TryDequeue(out buffer))
                {
                    buffer.Stream.Dispose();
                    byte[] bytes = null;
                    if (_pooledBuffers.TryDequeue(out bytes))
                        _bufferPool.Return(bytes);
                }
            }

            CheckBufferCount();
        }

    }
}
```

I then re-ran the exact same test and, well this is the result:

![NoError](https://github.com/MonoGame/MonoGame/assets/74532136/7037a3eb-f118-4551-b2af-23233a5c4c00)

And it's still running now without error.

**Comment by mrhelmut** [comment]
You're starting to be edgy, could you please stick to a factual tone and no subtext?

I didn't say that I didn't encounter a crash, but that I don't think it is a bug but rather an hardware limit based on your sample. On my computer it crashes right away with an XAudio2 error that a limit has been hit. Debugging it shows that it crashes exactly at 64+1 queued buffers, which is the limit of queued buffers into an XAudio2 voice. Therefore, not a bug but a misuse.

The problem that I have with this discussion, is that you seem to be "guess-changing" code (sorry if you're not, but that's the impression you're leaving). This is why I'm trying to make you explain your precise use-case and how you implemented it. The reason why I'd like to figure out the root cause is because applying fixes because "it works for me" is the best way to plant a tree that will later hide a forest of other issues.

-------

So, I dug into this deeper. It's a threading issue between ```SubmitBuffer()``` and ```OnBufferEnd()``` because XAudio2 callbacks (which are calling ```OnBufferEnd()```) are sent to C# from a different thread than the thread ```SubmitBuffer()``` was called from. 

This only happens when you submit too many too small buffers too close to each other. It starts making ```_queuedBuffers``` and ```_pooledBuffers``` to be out-of-sync and at some point it crashes because ```OnBufferEnd()``` tries to dequeue an inexistant buffer.

Your changes are not fixing this issue, just barely hiding it. ```ConcurrentQueue``` are useless because the two queues aren't part of the same lock (it just makes the imbalance less likely but still possible). Your changes in disposal code are patches over the actual crash, but don't fix the imbalance and still let the queues run loose, which can cascade into other unforeseen memory issues.

The actual fix to this would be to add a single lock around ```PlatformSubmitBuffer()```, ```PlatformDispose()```, and ```OnBufferEnd()```, like this:

```csharp
        private void PlatformSubmitBuffer(byte[] buffer, int offset, int count)
        {
            // XAudio2 callbacks are sent from a different thread,
            // so we need to lock our queues
            lock (_queuedBuffers)
            {
                if (_queuedBuffers.Count >= 64)
                    throw new ArgumentException("You have reached the limit of 64 queued buffers of XAudio2. Please use PendingBufferCount to avoid submitting that many buffers.");

                // we need to copy so datastream does not pin the buffer that the user might modify later
                byte[] pooledBuffer;
                pooledBuffer = _bufferPool.Get(count);
                _pooledBuffers.Enqueue(pooledBuffer);
                Buffer.BlockCopy(buffer, offset, pooledBuffer, 0, count);

                var stream = DataStream.Create(pooledBuffer, true, false, 0, true);
                var audioBuffer = new AudioBuffer(stream);
                audioBuffer.AudioBytes = count;

                _voice.SubmitSourceBuffer(audioBuffer, null);
                _queuedBuffers.Enqueue(audioBuffer);
            }
        }

        private void PlatformDispose(bool disposing)
        {
            if (disposing)
            {
                lock (_queuedBuffers)
                {
                    while (_queuedBuffers.Count > 0)
                    {
                        var buffer = _queuedBuffers.Dequeue();
                        buffer.Stream.Dispose();
                        _bufferPool.Return(_pooledBuffers.Dequeue());
                    }
                }
            }
            // _voice is disposed by SoundEffectInstance.PlatformDispose
        }

        private void OnBufferEnd(IntPtr obj)
        {
            // XAudio2 callbacks are sent from a different thread,
            // so we need to lock our queues
            lock (_queuedBuffers)
            {
                // Release the buffer
                if (_queuedBuffers.Count > 0)
                {
                    var buffer = _queuedBuffers.Dequeue();
                    buffer.Stream.Dispose();
                    _bufferPool.Return(_pooledBuffers.Dequeue());
                }

                CheckBufferCount();
            }
        }
```

I've submitted those changes to #8122 and added an exception when more than 64 buffers are submitted.
